File: src/psilox/utils/Time.java
package psilox.utils;

public class Time {

	public static final long SECOND = 1000000000;
	public static final long MILLIS = 1000000;
	public static final long MICROS = 1000;
	
	public static final long MINUTE = 60 * SECOND;
	public static final long HOUR = 60 * MINUTE;
	public static final long DAY = 24 * HOUR;
	
	private static long profile;
	
	public static void profile() {
		profile = Time.now();
	}
	
	public static long now() {
		return System.nanoTime();
	}
	
	public static long since() {
		return since(profile);
	}
	
	public static long since(long time) {
		return System.nanoTime() - time;
	}

	public static double flipFlop(long offset, long duration) {
		double now = (now() + offset) % duration;
		if(now < duration / 2) {
			return now / (duration / 2);
		} else {
			return 1 - (now - duration / 2) / (duration / 2);
		}
	}
	
	public static int flipFlop(int max, long offset, long duration) {
		return (int) (flipFlop(offset, duration) * max);
	}
	
}

File: src/psilox/utility/Time.kt
package psilox.utility

object Time {

    val NANO = 1L
    val MICRO = NANO * 1000
    val MILLI = MICRO * 1000
    val SECOND = MILLI * 1000
    val MINUTE = SECOND * 60
    val HOUR = MINUTE * 60
    val DAY = HOUR * 24
    val YEAR = DAY * 365

    var now = 0L
        get() = System.nanoTime()
        private set

    private var profile = now

    fun since(time: Long = profile): Long = now - time
    fun sinceF(time: Long = profile, unit: Long = 1): Float = since(time) / unit.toFloat()

    fun profile() {
        profile = now
    }

    infix fun Number.per(x: Long): Long = x / this.toLong()
    infix fun Number.elapsed(x: Long): Boolean = since(this.toLong()) >= x

}